contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,434 @@
+
+  // Last error encountered during detach operation, if any.
+  // +optional
+  // This field must only be set by the entity completing the detach
+  // operation, i.e. the external-attacher.
+  DetachError VolumeError `json:"detachError,omitempty" protobuf:"bytes,3,opt,name=detachError,casttype=VolumeError"`
+}
+
+type VolumeError struct {
+  // Time the error was encountered.
+  // +optional
+  Time metav1.Time `json:"time,omitempty" protobuf:"bytes,1,opt,name=time"`
+
+  // String capturing the error encountered during Attach operation.
+  // This string maybe logged, so it should not contain sensitive
+  // information.
+  // +optional
+  Message string `json:",message,omitempty" protobuf:"bytes,2,opt,name=message"`
+}
+```
+
+### Kubernetes In-Tree CSI Volume Plugin
+
+A new in-tree Kubernetes CSI Volume plugin will contain all the logic required for Kubernetes to communicate with an arbitrary, out-of-tree, third-party CSI compatible volume driver.
+
+The existing Kubernetes volume components (attach/detach controller, PVC/PV controller, Kubelet volume manager) will handle the lifecycle of the CSI volume plugin operations (everything from triggering volume provisioning/deleting, attaching/detaching, and mounting/unmounting) just as they do for existing in-tree volume plugins.
+
+#### Proposed API
+
+A new `CSIPersistentVolumeSource` object will be added to the Kubernetes API. It will be part of the existing `PersistentVolumeSource` objects and thus can be used only via PersistentVolumes. For now we do not consider allowing CSI volumes directly from Pods without PersistentVolumeClaim.
+
+```GO
+type CSIPersistentVolumeSource struct {
+  // Driver is the name of the driver to use for this volume.
+  // Required.
+  Driver string `json:"driver" protobuf:"bytes,1,opt,name=driver"`
+
+  // VolumeHandle is the unique volume name returned by the CSI volume
+  // plugin’s CreateVolume to refer to the volume on all subsequent calls.
+  VolumeHandle string `json:"volumeHandle" protobuf:"bytes,2,opt,name=volumeHandle"`
+
+  // Optional: MountSecretRef is a reference to the secret object containing
+  // sensitive information to pass to the CSI driver during NodePublish.
+  // This may be empty if no secret is required. If the secret object contains
+  // more than one secret, all secrets are passed.
+  // +optional
+  MountSecretRef *SecretReference `json:"mountSecretRef,omitempty" protobuf:"bytes,3,opt,name=mountSecretRef"`
+
+  // Optional: AttachSecretRef is a reference to the secret object containing
+  // sensitive information to pass to the CSI driver during ControllerPublish.
+  // This may be empty if no secret is required. If the secret object contains
+  // more than one secret, all secrets are passed.
+  // +optional
+  AttachSecretRef *SecretReference `json:"attachSecretRef,omitempty" protobuf:"bytes,4,opt,name=attachSecretRef"`
```

Comment:
The node authorizer will not allow kubelets to access this secret

Reply:
Kubelet will not need access to AttachSecrets